src/pybel/manager/cache_manager.py

```diff
@@ -637,18 +637,31 @@ def has_name_version(self, name, version):
         """
         return self.session.query(exists().where(and_(Network.name == name, Network.version == version))).scalar()
 
-    @staticmethod
-    def iterate_singleton_edges_from_network(network):
+    def iterate_singleton_edges_from_network(self, network):
         """Gets all edges that only belong to the given network
 
         :type network: Network
         :rtype: iter[Edge]
         """
-        return (  # TODO implement with nested SQLAlchemy query for better speed
-            edge
-            for edge in network.edges
-            if edge.networks.count() == 1
+        ne1 = aliased(network_edge, name='ne1')
+        ne2 = aliased(network_edge, name='ne2')
+        singleton_edge_ids_for_network = (
+            self.session.query(ne1.c.edge_id)
+            .outerjoin(ne2, and_(
+                ne1.c.edge_id == ne2.c.edge_id,
+                ne1.c.network_id != ne2.c.network_id
+            ))
+            .filter(and_(
+                ne1.c.network_id == network.id,
+                ne2.c.edge_id == None
+            ))
+            .subquery('singleton_edge_ids_for_network')
         )
+        singleton_edges_for_network = (
+            self.session.query(Edge)
+            .filter(Edge.id.in_(singleton_edge_ids_for_network))
+        )
+        return iter(singleton_edges_for_network)
```

Comment:
Can this query be given directly to a .delete() statement, or does SQLAlchemy (or SQL) not allow that?

Reply:
Also, what's the purpose of the explicit `iter()`, since this is passed to a for loop later?